From 07f38b475098714104f6ec2a77e5e9101809dfa1 Mon Sep 17 00:00:00 2001
From: Huiwen He <hehuiwen@kylinos.cn>
Date: Wed, 18 Feb 2026 00:34:30 +0800
Subject: [PATCH v2 05/12] smb/client: use binary search for NT status to DOS
 mapping

The ntstatus_to_dos_map[] table is sorted now. Replace the linear search
with binary search to improve lookup performance.

Also remove the sentinel entry as it is no longer needed with ARRAY_SIZE().

Signed-off-by: Huiwen He <hehuiwen@kylinos.cn>
Reviewed-by: ChenXiaoSong <chenxiaosong@kylinos.cn>
---
 fs/smb/client/smb1maperror.c | 32 +++++++++++++++++++-------------
 1 file changed, 19 insertions(+), 13 deletions(-)

diff --git a/fs/smb/client/smb1maperror.c b/fs/smb/client/smb1maperror.c
index 0f0ffe0b6075..818d1a88f11c 100644
--- a/fs/smb/client/smb1maperror.c
+++ b/fs/smb/client/smb1maperror.c
@@ -107,6 +107,19 @@ static const struct smb_to_posix_error mapping_table_ERRSRV[] = {
 /*****************************************************************************
  *convert a NT status code to a dos class/code
  *****************************************************************************/
+
+static __always_inline int ntstatus_to_dos_cmp(const void *_key, const void *_pivot)
+{
+	__u32 key = *(__u32 *)_key;
+	const struct ntstatus_to_dos_err *pivot = _pivot;
+
+	if (key < pivot->ntstatus)
+		return -1;
+	if (key > pivot->ntstatus)
+		return 1;
+	return 0;
+}
+
 /* NT status -> dos error map */
 static const struct ntstatus_to_dos_err ntstatus_to_dos_map[] = {
 /*
@@ -114,22 +127,15 @@ static const struct ntstatus_to_dos_err ntstatus_to_dos_map[] = {
  * sorted by NT status code (ascending).
  */
 #include "smb1_mapping_table.c"
-	{0, 0, 0, NULL}
 };
 
 static const struct ntstatus_to_dos_err *
-ntstatus_to_dos(__u32 ntstatus)
+search_ntstatus_to_dos_map(__u32 ntstatus)
 {
-	int i;
-
-	/* Check nt_errstr to allow mapping of NT_STATUS_OK (0) */
-	for (i = 0; ntstatus_to_dos_map[i].nt_errstr; i++) {
-		if (ntstatus == ntstatus_to_dos_map[i].ntstatus) {
-			return &ntstatus_to_dos_map[i];
-		}
-	}
-
-	return NULL;
+	return __inline_bsearch(&ntstatus, ntstatus_to_dos_map,
+				ARRAY_SIZE(ntstatus_to_dos_map),
+				sizeof(struct ntstatus_to_dos_err),
+				ntstatus_to_dos_cmp);
 }
 
 int
@@ -151,7 +157,7 @@ map_smb_to_linux_error(char *buf, bool logErr)
 		/* translate the newer STATUS codes to old style SMB errors
 		 * and then to POSIX errors */
 		__u32 err = le32_to_cpu(smb->Status.CifsError);
-		const struct ntstatus_to_dos_err *map = ntstatus_to_dos(err);
+		const struct ntstatus_to_dos_err *map = search_ntstatus_to_dos_map(err);
 
 		if (map) {
 			if ((logErr && err != NT_STATUS_MORE_PROCESSING_REQUIRED) ||
-- 
2.52.0

